telethon/client/updates.py

```diff
@@ -288,6 +332,9 @@ async def catch_up(self: 'TelegramClient'):
     # the order that the updates arrive in to update the pts and date to
     # be always-increasing. There is also no need to make this async.
     def _handle_update(self: 'TelegramClient', update):
+        if self._state_cache.update_already_processed(update):
```

Comment:
This may break, because it is possible for updates to be out of order (not sure if it can happen if we don't handle updates produced by requests, though).

Reply:
It actually happens, will try to fix, but it turned out to be complicated

Reply:
Yes, it's tricky. To handle updates properly, they should be buffered for up to half a second, and if the gap wasn't filled by then, then it should automatically get difference. This is why I would have rather refactored the entire thing (…but that's a lot of work). After my implementation of https://github.com/Lonami/grammers was battle tested, since it's cleaner, it could maybe be ported over.

Reply:
@vanutp it does happen, I have a reliable reproducer

Reply:
I know, I already wrote about it above